Guide: Update `Optimizing Package Bundling` to include new Bundle Analyzer (#87246)
vercel/next.js: docs/01-app/02-guides/package-bundling.mdx

---

> title: Optmizing package bundling

Reviewer: ```suggestion
title: Optimizing package bundling
```

nav_title: Package Bundling
description: Learn how to analyze and optimize your application's server and client bundles with the Turbopack Bundle Analyzer, and the `@next/bundle-analyzer` plugin for Webpack.
related:
  description: Learn more about optimizing your application for production.
  links:
    - app/guides/production-checklist
---

> Large bundles can negatively impact your application's performance. <AppOnly>By default, Next.js automatically bundles packages imported inside Server Components and Route Handlers.</AppOnly><PagesOnly>By default, packages imported into your application are not bundled, which can be a source of performance issues if external packages are not pre-bundled, for example, if imported from a monorepo or `node_modules`.</PagesOnly>. In some cases, you may need to further optimize it.

Author: ```suggestion
Large bundles can negatively impact your application's performance. <AppOnly>By default, Next.js automatically bundles packages imported inside Server Components and Route Handlers.</AppOnly><PagesOnly>By default, packages imported into your application are not bundled, which can be a source of performance issues if external packages are not pre-bundled, for example, if imported from a monorepo or `node_modules`.</PagesOnly>. In some cases, you may need to further optimize client bundles.
```

Next.js provides two tools for analyzing bundles:

- [Turbopack Bundle Analyzer (experimental)](#turbopack-bundle-analyzer-experimental)
- [`@next/bundle-analyzer` for Webpack](#nextbundle-analyzer-for-webpack)

This guide will walk you through how to use each tool depending on your setup, and how to [optimize package imports](#optimizing-package-imports).

<AppOnly>

## Turbopack Bundle Analyzer (Experimental)

> > This experimental feature is available in v16.1 and above. You can share feedback on the [dedicated GitHub discussion](https://github.com/vercel/next.js/discussions/86731).

Reviewer: I am hosting a demo here: https://turbopack-bundle-analyzer-demo.vercel.sh/

we plan on keeping it roughly up to date, and will be linking from the blog post.

Author: Nice, will add 👍🏼

The Turbopack Bundle Analyzer is integrated with Turbopack's module graph. It understands your application's routes, server and client boundaries, CSS and asset imports.

### Step 1: Run the Turbopack Bundle Analyzer

To get started, run the following command and open up the interactive view in your browser.

```bash

> next experimental-analyze --serve

Reviewer 2: We have more of these I think, in other docs entries... but this one won't work as it is, it has to be `npx next...` -

Author: ```suggestion
npx next experimental-analyze --serve
```

> npx next experimental-analyze --serve

Reviewer: we are going to drop the `--serve` flag

Reviewer 3: ```suggestion
npx next experimental-analyze
```

related:
  description: Learn more about optimizing your application for production.
  links:
    - app/guides/production-checklist
---

Bundling is the process of combining your application code and its dependencies into optimized output files for the client and server. Smaller bundles load faster, reduce JavaScript execution time, improve [Core Web Vitals](https://web.dev/articles/vitals), and lower server cold start times.

Next.js automatically optimizes bundles by code splitting, tree-shaking, and other techniques. However, there are some cases where you may need to optimize your bundles manually.

There are two tools for analyzing your application's bundles:

- [Next.js Bundle Analyzer for Turbopack (experimental)](#nextjs-bundle-analyzer-experimental)
- [`@next/bundle-analyzer` plugin for Webpack](#nextbundle-analyzer-for-webpack)

This guide will walk you through how to use each tool and how to [optimize large bundles](#optimizing-large-bundles).

## Next.js Bundle Analyzer (Experimental)

> > Available in v16.1 and later. You can share feedback in the [dedicated GitHub discussion](https://github.com/vercel/next.js/discussions/86731) and view a [demo here](https://turbopack-bundle-analyzer-demo.vercel.sh/).

Reviewer 4: Links should always state the destination through URL or copy. In this instance i think the URl makes most sense
```suggestion
> Available in v16.1 and later. You can share feedback in the [dedicated GitHub discussion](https://github.com/vercel/next.js/discussions/86731) and view the demo at [turbopack-bundle-analyzer-demo.vercel.sh](https://turbopack-bundle-analyzer-demo.vercel.sh/).
```

The Next.js Bundle Analyzer is integrated with Turbopack's module graph. You can inspect server and client modules with precise import tracing, making it easier to find large dependencies.

### Step 1: Run the Turbopack Bundle Analyzer

To get started, run the following command and open up the interactive view in your browser.

```bash filename="Terminal"
npx next experimental-analyze
```

### Step 2: Filter and inspect modules

Within the UI, you can filter by route, environment (client or server), and type (JavaScript, CSS, JSON), or search by file:

<Video
  caption="Next.js bundle analyzer UI walkthrough"
  src="/videos/bundle-analyzer.mp4"
  width={1708}
  height={1080}
/>

### Step 3: Trace modules with import chains

The treemap shows each module as a rectangle. Where the size of the module is represented by the area of the rectangle.

> Click a module to see its size, inspect its full import chain and see exactly where it used in your application:

Reviewer 2: ````suggestion
Click a module to see its size, inspect its full import chain and see exactly where it’s used in your application:
````

<Image
  caption="Next.js Bundle Analyzer import chain view"
  srcLight="/docs/light/bundle-analyzer.png"
  srcDark="/docs/dark/bundle-analyzer.png"
  width={1600}
  height={874}
/>

### Step 4: Write output to disk for sharing or diffing

> If you want to share the analysis with teammates or compare bundle sizes before and after optimizations, you can save the analysis as a static file:

Reviewer 2: If one had just run the command, it'll also produce these files, but w/ a sever. 

With --output, you can skip the server, for example if your intention is to generate the files to use them elsewhere, upload to an internal site, etc.

            <code>
              {tokens.map((line, i) => (
                <div key={i} {...getLineProps({ line })}>
                  {line.map((token, key) => (
                    <span key={key} {...getTokenProps({ token })} />
                  ))}
                </div>
              ))}
            </code>
          </pre>
        )}
      </Highlight>
    </article>
  )
}
```

This increases bundle size because the client must download and execute the highlighting library, even though the result is static HTML.

Instead, move the highlighting logic to a Server Component and render the final HTML on the server. The client will only receive the rendered markup.

```tsx filename="app/blog/[slug]/page.tsx"
import { codeToHtml } from 'shiki'

export default async function Page() {
  const code = `export function hello() {
    console.log("hi")
  }`

  // The Shiki package runs on the server and is never bundled for the client.
  const highlightedHtml = await codeToHtml(code, {
    lang: 'tsx',
    theme: 'github-dark',
  })

  return (
    <article>
      <h1>{post.title}</h1>

      {/* Client receives plain markup */}
      <pre>

>         <code dangerouslySetInnerHTML={{ __html: highlightedHtml }} />

Reviewer 2: Nit: not related to the bundle analyzer, but is `dangerouslySetInnerHTML` with shiki always safe?

Author: Yeah, I know, not the best, but used in the [Shiki docs](https://shiki.style/packages/next#:~:text=return%20%3Cdiv%20dangerouslySetInnerHTML%3D%7B%7B%20__html%3A%20out%20%7D%7D%20/%3E).
